cli/azd/pkg/project/framework_service_external.go

```diff
@@ -0,0 +1,531 @@
+func (efs *ExternalFrameworkService) sendAndWaitWithProgress(
+	ctx context.Context,
+	req *azdext.FrameworkServiceMessage,
+	progress *async.Progress[ServiceProgress],
+	match func(*azdext.FrameworkServiceMessage) bool,
+) (*azdext.FrameworkServiceMessage, error) {
+	respChan := make(chan *azdext.FrameworkServiceMessage, 1)
+	efs.responseChans.Store(req.RequestId, respChan)
+	defer efs.responseChans.Delete(req.RequestId)
+
+	if err := efs.stream.Send(req); err != nil {
+		return nil, fmt.Errorf("failed to send request: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case resp := <-respChan:
+			if resp == nil {
+				return nil, fmt.Errorf("stream closed")
+			}
+
+			if resp.Error != nil {
+				return nil, fmt.Errorf("framework service error: %s", resp.Error.Message)
+			}
+
+			if progressMsg := resp.GetProgressMessage(); progressMsg != nil {
+				if progress != nil {
+					progress.SetProgress(ServiceProgress{
+						Message: progressMsg.Message,
+					})
+				}
+				continue // Wait for the actual response
+			}
+
+			if !match(resp) {
+				return nil, fmt.Errorf("received unexpected response type")
+			}
+
+			return resp, nil
+		}
+	}
+}
+
+// goroutine to receive and dispatch responses
+func (efs *ExternalFrameworkService) startResponseDispatcher() {
+	go func() {
+		for {
+			resp, err := efs.stream.Recv()
+			if err != nil {
+				// propagate error to all waiting calls
+				efs.responseChans.Range(func(key, value any) bool {
+					ch := value.(chan *azdext.FrameworkServiceMessage)
+					close(ch)
+					return true
+				})
+				return
+			}
+			if ch, ok := efs.responseChans.Load(resp.RequestId); ok {
+
+				ch.(chan *azdext.FrameworkServiceMessage) <- resp
+			} else {
+				log.Printf("No response channel found for RequestId: %s", resp.RequestId)
+			}
+		}
+	}()
+}
+
+func (efs *ExternalFrameworkService) wireConsole() func() {
+	stdOut := efs.extension.StdOut()
+	stdErr := efs.extension.StdErr()
+	stdOut.AddWriter(efs.console.Handles().Stdout)
+	stdErr.AddWriter(efs.console.Handles().Stderr)
+
+	return func() {
+		stdOut.RemoveWriter(efs.console.Handles().Stdout)
+		stdErr.RemoveWriter(efs.console.Handles().Stderr)
+	}
+}
+
+// Convert ServiceConfig to proto message
+func (efs *ExternalFrameworkService) toProtoServiceConfig(serviceConfig *ServiceConfig) (*azdext.ServiceConfig, error) {
+	if serviceConfig == nil {
+		return nil, nil
+	}
+
+	image, err := serviceConfig.Image.Envsubst(func(name string) string {
+		return "" // Use empty resolver for now, similar to service_target_external.go
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve image value: %w", err)
+	}
+
+	return &azdext.ServiceConfig{
+		Name:         serviceConfig.Name,
+		RelativePath: serviceConfig.RelativePath,
+		Host:         string(serviceConfig.Host),
+		Language:     string(serviceConfig.Language),
+		OutputPath:   serviceConfig.OutputPath,
+		Image:        image,
+	}, nil
+}
```

Comment:
After we get this in we should have enough examples to see how we can generalize some of these helper/transfomer functions into a base class that each provider can leverage.